### ISSUE TITLE
Snapshots not creating

### ISSUE THREAD

**Issue by saward** [miracle2k/k8s-snapshots#27]
Hi,

I'm trying v1.0.1, using manual config, but no snapshots are created.  Here's my latest config (testing snapshotting the only disk I have thus far):

```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: k8s-snapshots
  labels:
    app: k8s-snapshots
spec:
  replicas: 1
  selector:
    matchLabels:
      app: k8s-snapshots
  template:
    metadata:
      labels:
        app: k8s-snapshots
    spec:
      containers:
        - image: elsdoerfer/k8s-snapshots:v1.0.1
          name: k8s-snapshots
          env:
          - name: GCLOUD_PROJECT
            value: manse-cloud
          - name: VOLUMES
            value: "kube-cert-manager"
          - name: VOLUME_KUBE_CERT_MANAGER_DELTAS
            value: "PT1H P1D P7D"
          - name: VOLUME_KUBE_CERT_MANAGER_ZONE
            value: "australia-southeast1-c"
          - name: LOG_LEVEL
            value: "DEBUG"
          - name: GCLOUD_JSON_KEYFILE_STRING
            valueFrom:
              secretKeyRef:
                name: snapshots-service-account
                key: snapshots-service-account.json

```

Here's the logs after an hour or two:

```
2017-08-21T00:12:42.233421Z rule.from-config               [k8s_snapshots.config] deltas_str=PT1H P1D P7D message=rule.from-config rule=Rule(name='kube-cert-manager', deltas=[datetime.timedelta(0, 3600), datetime.timedelta(1), datetime.timedelta(7)], gce_disk='kube-cert-manager', gce_disk_zone='australia-southeast1-c', source=None) severity=INFO volume_name=kube-cert-manager zone=australia-southeast1-c
2017-08-21T00:12:42.237344Z foreign                        [asyncio] message=Using selector: EpollSelector severity=DEBUG
2017-08-21T00:12:42.238683Z Gathering tasks                [k8s_snapshots.core] message=Gathering tasks severity=DEBUG tasks=[<Task pending coro=<scheduler() running at /usr/local/lib/python3.6/site-packages/k8s_snapshots-0.0.0-py3.6.egg/k8s_snapshots/core.py:271>>, <Task pending coro=<backuper() running at /usr/local/lib/python3.6/site-packages/k8s_snapshots-0.0.0-py3.6.egg/k8s_snapshots/core.py:289>>]
2017-08-21T00:12:42.239732Z scheduler.start                [k8s_snapshots.core] message=scheduler.start severity=DEBUG
2017-08-21T00:12:42.240466Z watch_schedule.start           [k8s_snapshots.core] message=watch_schedule.start severity=DEBUG
2017-08-21T00:12:42.241388Z backuper.start                 [k8s_snapshots.core] message=backuper.start severity=DEBUG
2017-08-21T00:12:42.242186Z rule.heartbeat                 [k8s_snapshots.core] message=rule.heartbeat rules=None severity=INFO
2017-08-21T00:12:42.243085Z volume-events.watch            [k8s_snapshots.core] message=volume-events.watch severity=DEBUG
2017-08-21T00:12:42.245651Z foreign                        [googleapiclient.discovery] message=URL being requested: GET https://www.googleapis.com/discovery/v1/apis/compute/v1/rest severity=INFO
2017-08-21T00:12:42.905296Z foreign                        [googleapiclient.discovery] message=URL being requested: GET https://www.googleapis.com/compute/v1/projects/manse-cloud/global/snapshots?filter=labels.created-by+eq+k8s-snapshots&alt=json severity=INFO
2017-08-21T00:12:42.906601Z foreign                        [oauth2client.transport] message=Attempting refresh to obtain initial access_token severity=INFO
2017-08-21T00:12:42.943009Z watch-resources.worker.start   [k8s_snapshots.kube] message=watch-resources.worker.start resource_type_name=PersistentVolume severity=DEBUG
2017-08-21T00:12:42.944380Z kube-config.from-service-account [k8s_snapshots.context] message=kube-config.from-service-account severity=INFO
2017-08-21T00:12:42.950273Z foreign                        [oauth2client.crypt] message=[<removed because I don't know if it's sensitive>] severity=DEBUG
2017-08-21T00:12:42.951193Z foreign                        [oauth2client.client] message=Refreshing access_token severity=INFO
2017-08-21T00:12:42.952846Z foreign                        [urllib3.connectionpool] message=Starting new HTTPS connection (1): 10.55.240.1 severity=DEBUG
2017-08-21T00:12:42.963228Z foreign                        [urllib3.connectionpool] message=https://10.55.240.1:443 "GET /api/v1/persistentvolumes?watch=true HTTP/1.1" 200 None severity=DEBUG
2017-08-21T00:12:44.012902Z watch_schedule.wait-for-both   [k8s_snapshots.core] message=watch_schedule.wait-for-both severity=DEBUG
2017-08-21T00:12:44.907999Z watch-resources.worker.start   [k8s_snapshots.kube] message=watch-resources.worker.start resource_type_name=PersistentVolumeClaim severity=DEBUG
2017-08-21T00:12:44.910447Z foreign                        [urllib3.connectionpool] message=Starting new HTTPS connection (1): 10.55.240.1 severity=DEBUG
2017-08-21T00:12:44.921444Z foreign                        [urllib3.connectionpool] message=https://10.55.240.1:443 "GET /api/v1/namespaces/default/persistentvolumeclaims?watch=true HTTP/1.1" 200 None severity=DEBUG
2017-08-21T00:22:42.244870Z rule.heartbeat                 [k8s_snapshots.core] message=rule.heartbeat rules=None severity=INFO
2017-08-21T00:32:42.246586Z rule.heartbeat                 [k8s_snapshots.core] message=rule.heartbeat rules=None severity=INFO                                                                                                                                                   
2017-08-21T00:42:42.247750Z rule.heartbeat                 [k8s_snapshots.core] message=rule.heartbeat rules=None severity=INFO                                                                                                                                                   
2017-08-21T00:47:36.623109Z watch-resources.worker.finalized [k8s_snapshots.kube] message=watch-resources.worker.finalized resource_type_name=PersistentVolume severity=DEBUG                                                                                                     
2017-08-21T00:47:36.715707Z watch-resources.done           [k8s_snapshots.kube] message=watch-resources.done resource_type_name=PersistentVolume severity=DEBUG                                                                                                                   
2017-08-21T00:52:42.249679Z rule.heartbeat                 [k8s_snapshots.core] message=rule.heartbeat rules=None severity=INFO                                                                                                                                                   
2017-08-21T01:01:47.579088Z watch-resources.worker.finalized [k8s_snapshots.kube] message=watch-resources.worker.finalized resource_type_name=PersistentVolumeClaim severity=DEBUG                                                                                                
2017-08-21T01:01:47.602920Z watch-resources.done           [k8s_snapshots.kube] message=watch-resources.done resource_type_name=PersistentVolumeClaim severity=DEBUG                                                                                                              
2017-08-21T01:01:47.603821Z sync-get-rules.done            [k8s_snapshots.core] message=sync-get-rules.done severity=DEBUG                                                                                                                                                        
2017-08-21T01:01:47.604225Z get-rules.done                 [k8s_snapshots.core] message=get-rules.done severity=DEBUG                                                                                                                                                             
2017-08-21T01:02:42.251460Z rule.heartbeat                 [k8s_snapshots.core] message=rule.heartbeat rules=None severity=INFO                                                                                                                                                   
2017-08-21T01:12:42.253261Z rule.heartbeat                 [k8s_snapshots.core] message=rule.heartbeat rules=None severity=INFO                                                                                                                                                   
2017-08-21T01:22:42.254266Z rule.heartbeat                 [k8s_snapshots.core] message=rule.heartbeat rules=None severity=INFO                                                                                                                                                   
2017-08-21T01:32:42.255648Z rule.heartbeat                 [k8s_snapshots.core] message=rule.heartbeat rules=None severity=INFO 
```

Any idea why no snapshots are being created?  I've checked that the region is right, that the name of the disk is right, and the project is right, so I'm reasonably confident I haven't made a mistake there.  I've gone over the documentation a couple of times to see if I missed anything obvious.

Also, a further question -- older versions created snapshots immediately.  Should this one also be creating a snapshot immediately, or would it wait for an hour given the above deltas before creating the first?

Thanks!

**Comment by miracle2k** [comment]
Sorry for the trouble!

Here is what I can gather from the logs:

2017-08-21T00:47:36.623109Z watch-resources.worker.finalized
2017-08-21T01:01:47.602920Z watch-resources.done
2017-08-21T01:01:47.603821Z sync-get-rules.done

At this point it stops observing Kubernetes for changes. Not sure why that happens. 

Also:

2017-08-21T00:12:44.012902Z watch_schedule.wait-for-both

That seems to imply it  never receives both rules and existing snapshots.

If you have Python installed locally, you could try to debug this by adding a print() here:

https://github.com/miracle2k/k8s-snapshots/blob/71cc577d85c9b698d2032364d97998ea537e6a0f/k8s_snapshots/core.py#L256

And see what it outputs.

Having said that, in the current latest version I had to unfortunately remove the manual volumes functionality again. My plan is to replace it with a ThirdPartyResource (https://github.com/miracle2k/k8s-snapshots/issues/26). If you can wait a couple of days, I hope to find the time to add those, because I need that functionality myself.

**Comment by saward** [comment]
I'm quite happy to wait a few days for the updated method.  However, third party resources are being replaced by CustomResourceDefinition (CRD) in 1.8.  Both are available in 1.7:

https://kubernetes.io/docs/tasks/access-kubernetes-api/migrate-third-party-resource/

**Comment by miracle2k** [comment]
Try https://github.com/miracle2k/k8s-snapshots#manual-snapshot-rules and the "dev" image tag. It worked on my machine, but some more testing would certainly be good.

**Comment by saward** [comment]
Great, I'll try that out within a few days I expect!

**Comment by saward** [comment]
A quick test indicates it's working.  If I create a new snapshot resource, do I need to restart the k8s-snapshots pod?  So far it seems like I do, but I didn't wait for much longer than 10 or 20 minutes to see.

**Comment by miracle2k** [comment]
You shouldn't have to restart, with one exception: If the pod is started before you created the thirdpartyresource (or crd), it will no be watching those resources; in this case, you'd have to restart. I should add this to the documentation.

**Comment by saward** [comment]
It appears to me that I need to restart the pod each time I add a new  snapshot.  At the very least, I've had to do this every time thus far.

**Comment by miracle2k** [comment]
Can you post the log of what k8s-snapshots does when you add a new SnapshotRule resource?

**Comment by saward** [comment]
Just added the following:

```
apiVersion: "k8s-snapshots.elsdoerfer.com/v1"
kind: SnapshotRule
metadata:
  name: faketest
spec:
  deltas: PT1H P1D P30D
  backend: google
  disk:
     zone: australia-southeast1-c
     name: faketest
```

via 'kubectl create -f test.yaml'.

There was no output/change in the k8s logs following the creation of that rule, so nothing to share.  It does show up in the list if I run 'kubectl get sr'.

For the previous rule I created, I searched a large section of the logs, and found no mention to the rule name, even after it had had time to r egsiter, so I don't think that it will appear later.

Image I'm using: elsdoerfer/k8s-snapshots:dev
